### PR TITLE
Add the first iteration of the Ja Rule testing tool.

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    m_out_in_progress(false),
+    m_in_transfer(libusb_alloc_transfer(0)),
+    m_in_in_progress(false) {
+}
+
+OpenLightingDevice::~OpenLightingDevice() {
+  {
+    MutexLocker locker(&m_out_mutex);
+    if (m_out_in_progress) {
+      OLA_INFO << "cancel m_out_transfer";
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: SPaG Cancel

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+      libusb_cancel_transfer(m_out_transfer);
+    }
+  }
+
+  {
+    MutexLocker locker(&m_in_mutex);
+    if (m_in_in_progress) {
+      OLA_INFO << "cancel m_in_transfer";
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: SPaG Cancel

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+  MutexLocker locker(&m_out_mutex);
+  if (m_out_in_progress) {
+    OLA_WARN << "TX in progress, dropping message";
+    return false;
+  }
+
+  unsigned int offset = 0;
+  m_out_buffer[0] = SOF_IDENTIFIER;
+  m_out_buffer[1] = static_cast<uint8_t>(command & 0xff);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: You can use our SplitUInt16 here.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+  m_out_buffer[2] = static_cast<uint8_t>(command >> 8);
+  m_out_buffer[3] = static_cast<uint8_t>(size & 0xff);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: And here.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+                            kTimeout);
+
+  Clock clock;
+  clock.CurrentTime(&m_out_sent_time);
+  OLA_INFO << "TX: Sending " << offset << " bytes";
+
+  int r = libusb_submit_transfer(m_out_transfer);
+
+  if (r) {
+    OLA_WARN << "Failed to submit out transfer: " << libusb_error_name(r);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: This needs to be LibUsbAdaptor::ErrorCodeToString.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    m_out_in_progress = false;
+  }
+}
+
+void OpenLightingDevice::_InTransferComplete() {
+  TimeStamp now;
+  Clock clock;
+  clock.CurrentTime(&now);
+  OLA_INFO << "Command completed in " << (now - m_out_sent_time)
+           << ", status is " << libusb_error_name(m_in_transfer->status);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: And again.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+  {
+    MutexLocker locker(&m_out_mutex);
+    m_in_in_progress = false;
+  }
+}
+
+bool OpenLightingDevice::SubmitInTransfer() {
+  MutexLocker locker(&m_in_mutex);
+  if (m_in_in_progress) {
+    OLA_WARN << "Reading already pending";
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: SPaG Read

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                            rx_buffer, IN_BUFFER_SIZE,
+                            InTransferCompleteHandler,
+                            static_cast<void*>(this),
+                            kTimeout);
+
+  Clock clock;
+  clock.CurrentTime(&m_send_in_time);
+  int r = libusb_submit_transfer(m_in_transfer);
+  if (r) {
+    OLA_WARN << "Failed to submit input transfer: " << libusb_error_name(r);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: And again.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    explicit ArrayDeleter(const uint8_t* data) : m_data(data) {}
+    ~ArrayDeleter() { delete[] m_data; }
+   private:
+    const uint8_t* m_data;
+  };
+
+  ArrayDeleter deleter(data);
+
+  // Right now we assume that the device only sends a single message at a time.
+  // If this ever change from a message model to more of a stream model we'll
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: SPaG changes

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+  }
+
+  if (size < MIN_RESPONSE_SIZE) {
+    OLA_WARN << "Response was too small, " << size << " bytes, min was "
+             << MIN_RESPONSE_SIZE;
+    return;
+  }
+
+  if (data[0] != SOF_IDENTIFIER) {
+    OLA_WARN << "SOF mismatch, was 0x" << ola::strings::ToHex(data[0]);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: This defaults to adding the 0x prefix, so you don't need to.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    return;
+  }
+
+  uint16_t command = data[1] + (data[2] << 8);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Use JoinUInt8

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+  uint16_t payload_size = data[3] + (data[4] << 8);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: And again.

File: tools/ja-rule/OpenLightingDevice.cpp [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+  uint8_t return_code = data[5];
+  uint8_t flags = data[6];
+
+  if (payload_size + MIN_RESPONSE_SIZE > size) {
+    OLA_WARN << "Message size of " << (payload_size + MIN_RESPONSE_SIZE)
+             << " is greater than rx size of " << size;
+    return;
+  }
+
+  ola::strings::FormatData(&std::cout, data, size);
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.cpp]: Does this really want to be cout?

File: tools/ja-rule/OpenLightingDevice.h [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    ECHO_COMMAND = 0x80,
+    TX_DMX = 0x81,
+    GET_LOG = 0x82,
+    GET_FLAGS = 0x83,
+    WRITE_LOG = 0x84
+  } Command;
+
+  /**
+   * @brief Create a new OpenLightingDevice.
+   * @param ss
```
Comment (peternewman) [inline tools/ja-rule/OpenLightingDevice.h]: Stop cheating!

File: tools/ja-rule/USBDeviceManager.cpp [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+  if (m_context) {
+    Stop();
+  }
+}
+
+bool USBDeviceManager::Start() {
+  int r = libusb_init(&m_context);
+
+  if (r < 0) {
+    OLA_WARN << "libusb_init() failed: " << libusb_error_name(r);
```
Comment (peternewman) [inline tools/ja-rule/USBDeviceManager.cpp]: Again.

File: tools/ja-rule/USBDeviceManager.cpp [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+
+void USBDeviceManager::HotPlugEvent(struct libusb_device *usb_device,
+                                    libusb_hotplug_event event) {
+  // See the comments under libusb_hotplug_register_callback in the libusb
+  // documentation. There are a number of caveats.
+  struct libusb_device_descriptor descriptor;
+  libusb_get_device_descriptor(usb_device, &descriptor);
+
+  if (descriptor.idVendor != kVendorId || descriptor.idProduct != kProductId) {
+    return;
```
Comment (peternewman) [inline tools/ja-rule/USBDeviceManager.cpp]: I think we should OLA_DEBUG the detected vendor and product, like we do with the USB plugin.

File: tools/ja-rule/USBDeviceManager.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * @brief Indicates a device has been added or removed
+   */
+  typedef ola::Callback2<void, EventType, OpenLightingDevice*>
+      NotificationCallback;
+
+  /**
+   * @brief Create a new USBDeviceManager.
+   * @param ss The executor to run the notification_cb on.
+   * @param notification_cb the callback to run when the device is added or
```
Comment (peternewman) [inline tools/ja-rule/USBDeviceManager.h]: SPaG The

File: tools/ja-rule/ja-rule.cpp [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+  }
+
+  DISALLOW_COPY_AND_ASSIGN(InputHandler);
+};
+
+/*
+ * Main.
+ */
+int main(int argc, char **argv) {
+  ola::AppInit(&argc, argv, "[ options ]", "Ja Rule Admin Tool");
```
Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: How come you're not using our flags?

**Comment by peternewman** [comment]
Are you really still calling it ja-rule?

**Comment by peternewman** [comment]
Looks generally good, just some mostly minor comments.

File: tools/Makefile.mk [hunk 19]
```diff
@@ -1,5 +1,6 @@
-include tools/ola_trigger/Makefile.mk
+include tools/ja-rule/Makefile.mk
```
Comment (peternewman) [inline tools/Makefile.mk]: This needs to be excluded on WIN32 as the StdIn handler it uses doesn't work for E1.33 so won't here either. It should also only be built if libusb is present.

**Comment by peternewman** [comment]
We should enable this on Travis too.

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 1 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Fixed

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 2 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Fixed

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 3 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 4 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 5 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 6 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 7 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Fixed.

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 8 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 9 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 10 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Removed

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 11 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 12 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Done

File: tools/ja-rule/OpenLightingDevice.cpp (hunk 13 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.cpp]: Yeah, it's just for me to debug right now. I'll remove it later - I added a TODO.

File: tools/ja-rule/OpenLightingDevice.h (hunk 14 above)
Comment (nomis52) [inline tools/ja-rule/OpenLightingDevice.h]: Fixed

File: tools/ja-rule/USBDeviceManager.cpp (hunk 15 above)
Comment (nomis52) [inline tools/ja-rule/USBDeviceManager.cpp]: Done

File: tools/ja-rule/USBDeviceManager.cpp (hunk 16 above)
Comment (nomis52) [inline tools/ja-rule/USBDeviceManager.cpp]: Added

File: tools/ja-rule/USBDeviceManager.h (hunk 17 above)
Comment (nomis52) [inline tools/ja-rule/USBDeviceManager.h]: Done

File: tools/ja-rule/ja-rule.cpp (hunk 18 above)
Comment (nomis52) [inline tools/ja-rule/ja-rule.cpp]: What flags? Right now there are no extra flags for this binary. AppInit takes care of parsing the flags.

Comment (peternewman) [inline tools/ja-rule/ja-rule.cpp]: Apologies, I'm not sure what I was thinking!

**Comment by peternewman** [comment]
You're calling it ja-rule as live too then @nomis52 . You've only enabled it on Windows and we should add it on Travis too.

**Comment by nomis52** [comment]
I fixed the Windows issue & enabled on travis.

**Pull request merged**